File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel


class DashboardResponse(BaseModel):
    total_incidents: int
    status_distribution: dict[str, int]
    severity_distribution: dict[str, int]
    category_distribution: dict[str, int]
    priority_distribution: dict[str, int]
    ai_status_distribution: dict[str, int]
    average_severity_confidence: float | None = None
    average_category_confidence: float | None = None


class TrendPoint(BaseModel):
    date: str
    count: int
    critical_count: int
    high_count: int
    medium_count: int
    low_count: int


class TrendResponse(BaseModel):
    trends: list[TrendPoint]


class HeatmapPoint(BaseModel):
    id: int
    latitude: float
    longitude: float
    severity: str
    category: str
    priority: str
    status: str
    created_at: datetime


class HeatmapResponse(BaseModel):
    incidents: list[HeatmapPoint]

File: frontend/src/api/analytics.ts
import api from "./client";

export interface DashboardAnalytics {
  total_incidents: number;
  status_distribution: Record<string, number>;
  severity_distribution: Record<string, number>;
  category_distribution: Record<string, number>;
  priority_distribution: Record<string, number>;
  ai_status_distribution: Record<string, number>;
  average_severity_confidence?: number | null;
  average_category_confidence?: number | null;
}

export interface TrendPoint {
  date: string;
  count: number;
  critical_count: number;
  high_count: number;
  medium_count: number;
  low_count: number;
}

export interface TrendResponse {
  trends: TrendPoint[];
}

export interface HeatmapPoint {
  id: number;
  latitude: number;
  longitude: number;
  severity: string;
  category: string;
  priority: string;
  status: string;
  created_at: string;
}

export interface HeatmapResponse {
  incidents: HeatmapPoint[];
}

export async function getDashboardAnalytics(): Promise<DashboardAnalytics> {
  const response = await api.get<DashboardAnalytics>(
    "/analytics/dashboard"
  );

  return response.data;
}

export async function getIncidentTrends(): Promise<TrendResponse> {
  const response = await api.get<TrendResponse>(
    "/analytics/trends"
  );

  return response.data;
}

export async function getIncidentHeatmap(): Promise<HeatmapResponse> {
  const response = await api.get<HeatmapResponse>(
    "/analytics/heatmap"
  );

  return response.data;
}